scripts/build_osm_network.py

```diff
@@ -41,823 +57,1576 @@
+    )
+    lines_all["s_nom"] = lines_all.apply(
+        lambda row: np.sqrt(3) * row["i_nom"] * row["voltage"] / 1e3, axis=1
+    )
 
-    # set substation_lv
-    set_lv_substations(buses)
+    buses_all["s_nom_sum0"] = lines_all.groupby("bus0")["s_nom"].sum()
+    buses_all["s_nom_sum0"] = buses_all["s_nom_sum0"].fillna(0)
+    buses_all["s_nom_sum1"] = lines_all.groupby("bus1")["s_nom"].sum()
+    buses_all["s_nom_sum1"] = buses_all["s_nom_sum1"].fillna(0)
+    buses_all["s_nom_sum"] = buses_all["s_nom_sum0"] + buses_all["s_nom_sum1"]
+    buses_all.drop(columns=["s_nom_sum0", "s_nom_sum1"], inplace=True)
+    buses_all["s_nom_sum"] = np.ceil(buses_all["s_nom_sum"]).astype(int)
 
-    # reset index
-    lines.reset_index(drop=True, inplace=True)
-    links.reset_index(drop=True, inplace=True)
+    buses_all.reset_index(inplace=True)
 
-    return lines, links, buses
+    return buses_all
 
 
-def _split_linestring_by_point(linestring, points):
+def _add_transformers(buses, geo_crs=GEO_CRS):
     """
-    Function to split a linestring geometry by multiple inner points.
+    Adds unique transformers between buses of different voltage levels at each station.
+
+    The function performs the following steps:
+    - Groups the buses by 'station_id' and identifies stations with buses of different voltage levels.
+    - Creates unique combinations of buses within each station and calculates the geometry for each transformer.
+    - Assigns unique transformer IDs based on station ID and voltage levels.
+    - Calculates the capacity of transformers based on the maximum capacity of connected buses.
 
-    Parameters
-    ----------
-    lstring : LineString
-        Linestring of the line to be split
-    points : list
-        List of points to split the linestring
+    Parameters:
+        - buses (GeoDataFrame): A GeoDataFrame containing bus information with columns including 'bus_id', 'station_id', 'voltage', and 'geometry'.
+        - geo_crs (CRS, optional): Coordinate reference system for the GeoDataFrame. Defaults to GEO_CRS.
 
-    Return
-    ------
-    list_lines : list
-        List of linestring to split the line
+    Returns:
+        - GeoDataFrame: A GeoDataFrame containing the added transformers with columns including 'transformer_id' and TRANSFORMERS_COLUMNS.
     """
+    buses_all = buses.copy().set_index("bus_id")
+    logger.info(
+        "Adding unique transformers between buses of different voltage levels at each station."
+    )
 
-    list_linestrings = [linestring]
+    all_transformers = gpd.GeoDataFrame(
+        columns=TRANSFORMERS_COLUMNS + ["transformer_id"], crs=geo_crs
+    )
 
-    for p in points:
-        # execute split to all lines and store results
-        temp_list = [split(l, p) for l in list_linestrings]
-        # nest all geometries
-        list_linestrings = [lstring for tval in temp_list for lstring in tval.geoms]
+    for g_name, g_value in buses_all.groupby("station_id"):
```

Comment:
*If this is slow,* could be converted to an apply function applied to each group.

Reply:
it is surprisingly fast, few seconds, so no need for now from my side :)